mozfldp/fed_avg_w_dp.py

```diff
@@ -0,0 +1,250 @@
+import random
+import numpy as np
+from sklearn.linear_model import SGDClassifier
+
+
+class FedAvgWithDpParams:
+    def __init__(
+        self,
+        num_users,
+        num_features,
+        num_labels,
+        num_rounds,
+        batch_size,
+        num_epochs,
+        weight_mod,
+        sensitivity,
+        noise_scale,
+        rand_seed,
+    ):
+        self.num_users = num_users
+        self.num_features = num_features
+        self.num_labels = num_labels
+        self.num_rounds = num_rounds
+        self.batch_size = batch_size
+        self.num_epochs = num_epochs
+        self.weight_mod = weight_mod  # w_hat
+        self.sensitivity = sensitivity
+        self.noise_scale = noise_scale
+        self.rand_seed = rand_seed
+
+
+def run_fed_avg_with_dp(prms, data):
+    """
+    Runs federated averaging with differential privacy
+    prms: The parameters needed to run (FedAvgWithDpParams)
+    data: Data to train on. In the format of: ([user_1_labels, user_2_labels, ...], [user_1_feats, user_2_feats, ...])
+    """
+
+    # Fix the seed for all rng calls
+    np.random.seed(prms.rand_seed)
+    random.seed(prms.rand_seed)
+
+    user_weights, weight_sum = _init_user_weights_and_weight_sum(
+        prms.num_users, prms.weight_mod
+    )
+    theta = None
+    theta_0 = _init_theta_from_moment_accountant(prms.num_features, prms.num_labels)
+    prev_theta = np.array(theta_0, copy=True)
+    user_sel_prob = 1.0 / prms.num_users
+    standard_dev = _calc_standard_dev(
+        prms.noise_scale, prms.sensitivity, user_sel_prob, weight_sum
+    )  # This feels weird being a constant...
```

Comment:
It's constant because it's related to the maximum theoretical contribution of any single client, which is determined by S (`prms.sensitivity`), as opposed to observed contributions.

Reply:
:+1: